Add support for findOneAndUpdate

paul-pagnan opened the pull request:
Hey Mikhail!

Another pull request for you. This time I have added support for the findOneAndUpdate method.

Previously, this library assumed that you would always save documents using the .save() method. So when updating using the findOneAndUpdate method, none of the save hooks are called and thus no event is logged. 

One thing that is missing here is the .update method, but this is not as common, so I have omitted it for now.

I will spend some time later tonight creating a Readme. Would be great to get this out into the public, it is extremely useful

Thanks,
Paul

EDIT: I have also added TS definitions

limenutt commented:
Nice implementation of the findOneAndUpdate logging!
Re TS definitions: I think it would be great to include typings that allow client implementations to type patched mongoose objects properly as well, e.g. being able to set actor on a query or document.
Have a look at the ones we used in KL.
[definitions.zip](https://github.com/teamintrepid/mongoose-event-logger/files/1799884/definitions.zip)

paul-pagnan commented:
Ahh nice one, I added the by function to my own model definitions for the time being, but create a patched version of the Document is a better idea. I will do that.

You also mentioned that there was a reason that you didn't implement findOneAndUpdate to begin with. Was there something I am missing?

limenutt commented:
Ah, the reason why we haven't implemented it was that with findOneAndUpdate as well as with the `update` you can specify just the fields to update, and therefore we won't be able to get a snapshot. You solved it by fetching the whole document in a hook, which now gives us the two versions to compare, derive the delta and the snapshot. There is a slight chance of inconsistency, when for example we fetched the document in a hook it has existed, but by the time `findOneAndUpdate` operation was sent to `mongodb`, it has already gone or doesn't match the find criteria. But such inconsistency is out of the scope of this plugin. It would be enough to mention that some methods, although supported, may log inconsistent events in some situations. Therefore developers would decide what methods to use based on their use case.

paul-pagnan commented:
Hmm, yes you raise a good point. There could definitely be race conditions here at the time we attempt to update the document and when we fetch the result to compare the delta. Another process might have already modified something. 
I agree though, I think that might be outside the scope of this plugin. Maybe we recommend using this plugin along with the lock plugin you also developed ;)

limenutt commented:
Yeah, I might publish that one soon as well :)
So, would you like to update this PR with the typings I've suggested or do you want to do it later?

paul-pagnan commented:
I might leave those types for a later date. Also need to write some documentation

I have just updated this pull request to fix the findOneAndUpdate implementation. I needed to bypass the validation before save when using findOneAndUpdate because it does not perform model validation. When we try to trigger the save hook with an object that does not pass the model validation,  it fails this validation on save and it would crash.

EDIT: I also added support for mongo db connection strings, rather than specifying the host, user and password individually (because we use replica sets, so the mongo connection string is more complex than just username, password, host and port)

limenutt commented on src/plugins/mongoose/logger/plugin.js:
```diff
@@ -241,10 +252,52 @@ export function plugin(mongooseInstance) {
+      result._klLoggerStaticContext = true;
+
+      // Skip validation when because findOneAndUpdate does not do model validation.
+      await result.save({ validateBeforeSave: false });
```
What is the point of saving the updated document `again` here?
Is it just to make `schema.post('save')` hook to fire and log the event? 
I think it would be better to abstract out the event logging logic and reuse it in both hooks.
